AtReconstruction/AtFitter/OpenKF/kalman_filter/TrackFitterUKFTest.cxx

```diff
@@ -0,0 +1,268 @@
+   ASSERT_NEAR(m_ukf.matP()(2, 1), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(2, 2), 0.15F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(2, 3), 0.0F, FLOAT_EPSILON);
+
+   ASSERT_NEAR(m_ukf.matP()(3, 0), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(3, 1), 0.05F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(3, 2), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(3, 3), 0.15F, FLOAT_EPSILON);
+}
+
+TEST_F(TrackFitterUKFExampleTest, PredictionAndCorrection)
+{
+   kf::Vector<DIM_X> x;
+   x << 2.0F, 1.0F, 0.0F, 0.0F;
+
+   kf::Matrix<DIM_X, DIM_X> P;
+   P << 0.01F, 0.0F, 0.0F, 0.0F, 0.0F, 0.01F, 0.0F, 0.0F, 0.0F, 0.0F, 0.05F, 0.0F, 0.0F, 0.0F, 0.0F, 0.05F;
+
+   kf::Matrix<DIM_V, DIM_V> Q;
+   Q << 0.05F, 0.0F, 0.0F, 0.0F, 0.0F, 0.05F, 0.0F, 0.0F, 0.0F, 0.0F, 0.1F, 0.0F, 0.0F, 0.0F, 0.0F, 0.1F;
+
+   kf::Matrix<DIM_N, DIM_N> R;
+   R << 0.01F, 0.0F, 0.0F, 0.01F;
+
+   kf::Vector<DIM_Z> z;
+   z << 2.5F, 0.05F;
+
+   m_ukf.vecX() = x;
+   m_ukf.matP() = P;
+
+   m_ukf.setCovarianceQ(Q);
+   m_ukf.setCovarianceR(R);
+
+   m_ukf.predictUKF(funcF);
+
+   // Expectation from the python results:
+   // =====================================
+   // x =
+   //     [2.0 1.0 0.0 0.0]
+   // P =
+   //     [[0.11  0.00  0.05  0.00]
+   //      [0.00  0.11  0.00  0.05]
+   //      [0.05  0.00  0.15  0.00]
+   //      [0.00  0.05  0.00  0.15]]
+
+   ASSERT_NEAR(m_ukf.vecX()[0], 2.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.vecX()[1], 1.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.vecX()[2], 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.vecX()[3], 0.0F, FLOAT_EPSILON);
+
+   ASSERT_NEAR(m_ukf.matP()(0, 0), 0.11F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(0, 1), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(0, 2), 0.05F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(0, 3), 0.0F, FLOAT_EPSILON);
+
+   ASSERT_NEAR(m_ukf.matP()(1, 0), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(1, 1), 0.11F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(1, 2), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(1, 3), 0.05F, FLOAT_EPSILON);
+
+   ASSERT_NEAR(m_ukf.matP()(2, 0), 0.05F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(2, 1), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(2, 2), 0.15F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(2, 3), 0.0F, FLOAT_EPSILON);
+
+   ASSERT_NEAR(m_ukf.matP()(3, 0), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(3, 1), 0.05F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(3, 2), 0.0F, FLOAT_EPSILON);
+   ASSERT_NEAR(m_ukf.matP()(3, 3), 0.15F, FLOAT_EPSILON);
+
+   m_ukf.correctUKF(funcH, z);
+
+   // Expectations from the python results:
+   // ======================================
+   // x =
+   //     [ 2.4758845   0.53327217  0.21649734 -0.21214576]
+   // P =
+   // [[ 0.01433114 -0.01026142  0.00651178 -0.00465059]
+   //  [-0.01026142  0.0295458  -0.0046378   0.01344241]
+   //  [ 0.00651178 -0.0046378   0.13023154 -0.00210188]
+   //  [-0.00465059  0.01344241 -0.00210188  0.1333886 ]]
```

Comment:
These values are from a python implementation of the correction I wrote. It's what I used to validate since I didn't have applicable test results from OpenKF. Python and cpp implementations agree so I think the math was implemented correctly